morph_tool/amira_converter.py

```diff
@@ -0,0 +1,172 @@
+        "@2": "node_label",
+        "@3": "edge",
+        "@4": "n_points",
+        "@5": "label",
+        "@6": "point",
+        "@7": "radius",
+    }
+    sections = defaultdict(list)
+    current_section = ""
+    for line in lines:
+        if line.startswith("@"):
+            current_section = line
+        elif not line:
+            current_section = ""
+        elif current_section:
+            sections[_converter[current_section]].append(line)
+    return sections
+
+
+def get_labels(lines, level=0):
+    """Get label mapping."""
+    label = None
+    line_id = 0
+    retry = False
+    _id = None
+    labels = {}
+    while line_id <= len(lines):
+        line = lines[line_id]
+
+        if "{" in line and not retry:
+            label = line.strip().split(" ")[0]
+
+        if "{" in line and "{" in lines[line_id + 1] and not retry:
+            _labels, n_lines = get_labels(lines[line_id + 1:], level + 1)
+            labels.update(_labels)
+            line_id += n_lines
+            retry = True
+        elif retry and "{" in line:
+            _labels, n_lines = get_labels(lines[line_id:], level + 1)
+            labels.update(_labels)
+            line_id += n_lines - 1
+            retry = True
+
+        if "Id" in line:
+            if line.endswith(","):
+                line = line[:-1]
+            _id = int(line.split(" ")[-1])
+
+        if "}" in line:
+            if _id is not None:
+                labels[_id] = label
+            break
+
+        line_id += 1
+
+    return labels, line_id + 1
+
+
+def create_dfs(sections, labels):
+    """Create dataframe with morphology data."""
+    point_id = 0
+    df = pd.DataFrame(columns=["u", "v", "label", "points", "diameters"])
+
+    us = []
+    vs = []
+    _labels = []
+    points = []
+    diameters = []
+    for edge, n_points, label_id in zip(sections["edge"], sections["n_points"], sections["label"]):
+        u, v = np.fromstring(edge, dtype=int, sep=" ")
+        point = []
+        diameter = []
+        for _ in range(int(n_points)):
+            point.append(np.fromstring(sections["point"][point_id], dtype=float, sep=" "))
+            diameter.append(2 * float(sections["radius"][point_id]))
+            point_id += 1
+        us.append(u)
+        vs.append(v)
+        _labels.append(labels[int(label_id)])
+        points.append([list(p) for p in point])
+        diameters.append(diameter)
+    df["u"] = us
+    df["v"] = vs
+    df["label"] = _labels
+    df["points"] = points
+    df["diameters"] = diameters
+    dfs = {}
+    for label, _df in df.groupby("label"):
+        dfs[label] = _df.sort_values(by="u")
+
+    return dfs
+
+
+def make_soma(dfs, morph):
+    """Make a soma."""
+    if "Soma" in dfs:
+        soma_points = []
+        soma_diameters = []
+        for gid in dfs["Soma"].index:
+            soma_points += [list(p) for p in dfs["Soma"].loc[gid, "points"]]
+            soma_diameters += list(dfs["Soma"].loc[gid, "diameters"])
```

Comment:
it will be too much work to 'flatten' the dfs, by adding a node per points instead of per bif, I don't want to do that just to skip a loop here ( this will need more loops before anyways)

Reply:
ok, no problem